core/lib/src/server.rs

```diff
@@ -342,6 +342,20 @@ impl Rocket<Orbit> {
         }
 
         error_!("No matching routes for {}.", request);
+        if request.route().is_none() {
+            // We failed to find a route which matches on path, query AND formats.
+            if self.router.matches_except_formats(request) {
+                // Tailor the error code to the interpretation of the request in question.
+                if request.method().supports_payload() {
+                    status = Status::UnsupportedMediaType;
+                } else {
+                    status = Status::NotAcceptable;
+                }
```

Comment:
I don’t think that returning `NotAcceptable` is correct here. This status code is about content negotiation via the request’s `Accept` header – e.g. the client wants application/json but the server can only deliver text/html. See https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/406

IMHO this branch should always return `UnsupportedMediaType` regardless of whether payload is supported.

Reply:
I don't quite follow your reasoning here. Rocket's route "format" specifier switches on exactly whether or not the method expects payload or not ( https://rocket.rs/v0.5/guide/requests/#format )
The logic here merely mirrors that same logic, and returns 406 if the route can only produce payload which the client doesn't want.
It would be wrong to return 415 for a GET route which can't return the desired content.

Reply:
My bad, I didn’t realize that this restriction had different meanings depending on the method. Your implementation is correct.